### PR TITLE
chore: write our own swagger bindings

### PR DESCRIPTION
**Pull request by rb-determined-ai**
try:

    cd bindings
    python bindings.py > b.py
    pyinstrument b.py
    mypy b.py

On my laptop, `pyinstrument b.py` shows:
 - 20ms for the python interpreter to read/compile the code
 - 10ms to actually import the module

## Commentary (optional)

Initial draft, passes mypy.  Not tested yet.

### CODE REVIEW

File: bindings/bindings.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+    def dump(self, val: Code) -> Code:
+        return f"{self.name}.from_json({val})"
+
+    def gen_def(self) -> Code:
+        out = [f"class {self.name}:"]
+        out += ["    def __init__("]
+        out += ["        self,"]
+        out += [f"        {k}: {v.annotation()}," for k, v in self.members.items()]
+        out += ["    ):"]
+        out += [f"        {k} = {k}" for k in self.members]
```
Comment (ioga) [inline bindings/bindings.py:225]:
```suggestion
        out += [f"        self.{k} = {k}" for k in self.members]
```

File: bindings/bindings.py [hunk 2]
```diff
@@ -0,0 +1,558 @@
+import abc
```
Comment (ioga) [inline bindings/bindings.py:1]: that's not used; this file itself is not under mypy so there's a few "linter issues" here and there.

File: bindings/bindings.py [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+class Enum(TypeDef):
+    def __init__(self, name, members):
+        self.name = name
+        self.members = members
+
+    def load(self, val: Code) -> Code:
+        return f"{val}.value"
+
+    def dump(self, val: Code) -> Code:
+        return f"{self.name}({val})"
```
Comment (ioga) [inline bindings/bindings.py:260]: I think these two are swapped

```suggestion
    def load(self, val: Code) -> Code:
        return f"{self.name}({val})"

    def dump(self, val: Code) -> Code:
        return f"{val}.value"
```

**Comment by hamidzr** [comment]
this sounds great, I'm working on a review. Would we want to also commit the generated code in this pr?

**Review by hamidzr — COMMENTED** [review]
I ran into some issues testing the generated code. after applying a patch and fixing a seemingly swapped load and dump

```
Traceback (most recent call last):
  File "/home/hmd/projects/da/determined/harness/determined/cli/cli.py", line 240, in main
    parsed_args.func(parsed_args)
  File "/home/hmd/projects/da/determined/harness/determined/cli/master.py", line 44, in get_master
    print(resp.to_json())
  File "/home/hmd/projects/da/determined/harness/determined/common/api/b.py", line 1523, in to_json
    "version": self.version,
AttributeError: 'v1GetMasterResponse' object has no attribute 'version'
Failed to fetch master info
```
[bindings.patch.txt](https://github.com/determined-ai/determined/files/7488572/bindings.patch.txt)

**Comment by rb-determined-ai** [comment]
I'll take a look when I get a chance

**Comment by hamidzr** [comment]
I found the issue, fixed it and added an example https://github.com/determined-ai/determined/compare/master...hamidzr:inhouse-bindings?expand=1

**Comment by hamidzr** [comment]
something else we might want camelCase to snake_case conversion for python. maybe not worth it.

**Comment by rb-determined-ai** [comment]
pushed my updates.  Looks like you have some fixes I don't.  I'm fine with you taking this over, I clearly don't have the bandwidth and you are blocked on it.

**Comment by rb-determined-ai** [comment]
> something else we might want camelCase to snake_case conversion for python. maybe not worth it.

See https://github.com/determined-ai/determined/blob/master/schemas/gen.py#L15.  If you choose to use that function, please just copy/paste it.  The cross-repo import would be much messier than a copy/paste I think.

**Comment by rb-determined-ai** [comment]
closing in favor of https://github.com/determined-ai/determined/pull/3361

**Pull request closed without merging**